server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -685,73 +840,56 @@ private static void parseNonDynamicArray(
+            autoEndEventSender = context.publishEvent(
+                new DocumentParserListener.Event.LeafArrayStart(fieldMapper, context.parent(), context.doc())
+            );
         }
 
-        // In synthetic source, if any array element requires storing its source as-is, it takes precedence over
-        // elements from regular source loading that are then skipped from the synthesized array source.
-        // To prevent this, we track that parsing sub-context is within array scope.
-        context = context.maybeCloneForArray(mapper);
+        try (var endSender = autoEndEventSender) {
+            // Check if we need to record the array source. This only applies to synthetic source.
+            if (context.canAddIgnoredField()) {
+                boolean copyToFieldHasValuesInDocument = context.isWithinCopyTo() == false && context.isCopyToDestinationField(fullPath);
+                if (copyToFieldHasValuesInDocument) {
+                    context = context.addIgnoredFieldFromContext(IgnoredSourceFieldMapper.NameValue.fromContext(context, fullPath, null));
+                } else if (mapper instanceof ObjectMapper objectMapper && (objectMapper.isEnabled() == false)) {
+                    // No need to call #addIgnoredFieldFromContext as both singleton and array instances of this object
+                    // get tracked through ignored source.
+                    // context.addIgnoredField(
+                    // IgnoredSourceFieldMapper.NameValue.fromContext(context, fullPath, context.encodeFlattenedToken())
+                    // );
+                    return;
+                }
+            }
 
-        XContentParser parser = context.parser();
-        XContentParser.Token token;
-        int elements = 0;
-        while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
-            if (token == XContentParser.Token.START_OBJECT) {
-                elements = Integer.MAX_VALUE;
-                parseObject(context, lastFieldName);
-            } else if (token == XContentParser.Token.START_ARRAY) {
-                elements = Integer.MAX_VALUE;
-                parseArray(context, lastFieldName);
-            } else if (token == XContentParser.Token.VALUE_NULL) {
-                elements++;
-                parseNullValue(context, lastFieldName);
-            } else if (token == null) {
-                throwEOFOnParseArray(arrayFieldName, context);
-            } else {
-                assert token.isValue();
-                elements++;
-                parseValue(context, lastFieldName);
+            // In synthetic source, if any array element requires storing its source as-is, it takes precedence over
+            // elements from regular source loading that are then skipped from the synthesized array source.
+            // To prevent this, we track that parsing sub-context is within array scope.
+            context = context.maybeCloneForArray(mapper);
```

Comment:
Is this still needed?

Reply:
We still need this for higher-level array tracking.